sycl/doc/extensions/proposed/sycl_ext_oneapi_graph_fusion.asciidoc

```diff
@@ -0,0 +1,644 @@
+the APIs, making sure users only need to familiarize themselves with a single
+API, but also provides additional advantages. 
+
+The graph proposal defines two APIs to create graphs: a proposal using a
+recording mechanism, similar to the initial kernel fusion proposal; and another
+one using explicit graph building. Thus, future users will be able to choose
+from two different mechanisms to construct the sequence of kernels to fuse. As
+there is an explicit step for finalization of graphs before being submitted for
+execution, fusion can happen in this step, which also eliminates many of the
+synchronization concerns that needed to be covered in the experimental kernel
+fusion proposal.
+
+The aim of this document is to propose a mechanism for users to request the
+fusion of two or more kernels in a SYCL graph into a single kernel **at
+runtime**. This requires the extension of the runtime with some sort of JIT
+compiler to allow for the fusion of kernel functions at runtime.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_GRAPH_FUSION` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to determine
+if the implementation supports this feature, or applications can test the
+macro's value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== API modifications
+
+==== Properties
+
+===== Graph Fusion Property
+
+The API for `command_graph<graph_state::modifiable>::finalize()` includes a
+`property_list` parameter. The following property, defined by this extension,
+can be added to the property list to indicate that the kernels in the
+command-graph should be fused. 
+
+```c++
+sycl::ext::oneapi::experimental::property::command_graph::perform_fusion 
+```
+
+The property is not prescriptive. Implementations are free to not perform fusion
+if it is not possible (see below section <<limitations, Limitations>>), fusion is not
+supported by the implementation, or the implementation decides not to perform
+fusion for other reasons. It is not an error if an implementation does not
+perform fusion even though the property is passed. 
+
+Implementations can provide a diagnostic message in case fusion was not
+performed through an implementation-specified mechanism, but are not required to
+do so.
+
+===== Barrier property
+
+The following property can be added to the `property_list` of the
+`command_graph<graph_state::modifiable>::finalize()` API.
+
+```c++
+sycl::ext::oneapi::experimental::property::command_graph::no_barriers
+```
+
+If the property list contains this property, no barriers are introduced between
+kernels in the fused kernel (see below section on synchronization on kernels). 
```

Comment:
I think this is too coarse-grained. 

I can see how it would be useful for simple cases, but couldn't there be cases where the developer wants to fuse a chain like `k1` -> `k2` -> `k3`, and they need a barrier between `k1` and `k2`, but not between `k2` and `k3`?

What about making the "no barriers" behavior the default instead, and requiring the developer to insert explicit barriers between pairs of kernels that will still require a barrier after fusion?  You could represent such a barrier using something similar to [sycl_ext_oneapi_enqueue_barrier](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_enqueue_barrier.asciidoc).

Reply:
> You could represent such a barrier using something similar to [sycl_ext_oneapi_enqueue_barrier](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_enqueue_barrier.asciidoc).

A note that current the `sycl_ext_oneapi_graph` spec currently forbids using `sycl_ext_oneapi_enqueue_barrier` in a graph node, but this is a restriction we'll relax in the near future given we're in the process of implementing it in our fork https://github.com/reble/llvm/pull/301

Reply:
> Another easy way is to add a `group_barrier` at the end/beginning of a kernel, because they do not get removed by fusion.

This seems dangerous to me.  I think it's quite likely that optimizing compilers may try to remove redundant (i.e. back-to-back) barriers or unused barriers as a performance optimization.  For your suggestion to work, we'd have to try and make sure that fusion always happens before any such optimizations.

> Therefore, we also wanted to make disablement of the barrier insertion an active decision by the user (through the property).
> 
> Would you still prefer the default to be changed?

I don't feel strongly about the default -- my concern is really about the granularity of control.

If the default is to insert work-group barriers between kernels, then I think there needs to be a mechanism for developers to express that a barrier is not required between a specific pair of kernels (i.e. without removing all barriers between all pairs of fused kernels).

If the default is not to insert work-group barriers between kernels, then I think there needs to be a mechanism for developers to express that a barrier is required between a specific pair of kernels.  I was leaning towards this just because it seemed we already had the necessary functionality (enqueueing a barrier).

Sorry for not being clear about that originally.

Reply:
Maybe the answer to fine-grained control is sub-graphs. If the user wants device-wide synchronization, they could simply create two sub-graphs (one for kernels before the barrier, one for kernels after the barrier), fuse each of them and then submit them as sub-graphs of a larger graph without fusion. 

This would allow for the fine-grained control we were looking for. WDYT?

Reply:
That works for me.  The idea of using sub-graphs also ties in nicely to the prescriptive/descriptive versions of the properties we've been discussing -- I can imagine cases where a developer expects (and thus prescribes) fusion within a few small sub-graphs consisting of a handful of kernels, but leaves the decision to fuse other things up to the implementation.

A non-normative note explaining that developers can use sub-graphs like this would be really helpful, I think.

Reply:
I changed the default behavior and introduced the `insert_barriers` property to trigger insertion of work-group barriers between every pair of kernels.